File: pkg/AMORE/src/MLPfactory.cpp
/*
 * MLPfactory.cpp
 *
 *  Created on: 14/07/2011
 *      Author: mcasl
 */

#include "package.h"
#include "classHeaders/Connection.h"
#include "classHeaders/Neuron.h"
#include "classHeaders/SimpleNeuron.h"
#include "classHeaders/Container.h"
#include "classHeaders/SimpleContainer.h"
#include "classHeaders/NeuralNetwork.h"
#include "classHeaders/SimpleNetwork.h"
#include "classHeaders/NeuralCreator.h"
#include "classHeaders/SimpleNeuralCreator.h"
#include "classHeaders/predictBehavior.h"
#include "classHeaders/MLPbehavior.h"
#include "classHeaders/Iterator.h"
#include "classHeaders/NeuralFactory.h"
#include "classHeaders/MLPfactory.h"

//=========================================================================================================


ConPtr
MLPfactory::makeCon(Neuron& neuron, double weight)
{
  ConPtr conPtr(new Con(neuron, weight));
  return conPtr;
}

ConContainerPtr
MLPfactory::makeConContainer()
{
  ConContainerPtr conContainerPtr(new SimpleContainer<ConPtr> );
  return conContainerPtr;
}

PredictBehaviorPtr
MLPfactory::makePredictBehavior(NeuronPtr neuronPtr)
{
  PredictBehaviorPtr predictBehaviorPtr(new MLPbehavior(neuronPtr));
  return predictBehaviorPtr;
}

NeuronPtr
MLPfactory::makeNeuron(Handler Id)
{
  NeuronPtr neuronPtr(new SimpleNeuron(*this));
  neuronPtr->setId(Id);
  neuronPtr->setPredictBehavior(makePredictBehavior(neuronPtr));
  neuronPtr->setActivationFunction(
      makeActivationFunction(neuronPtr,
          Rcpp::XPtr<CppFunctionPointer>(new CppFunctionPointer(&default_f0)),
          Rcpp::XPtr<CppFunctionPointer>(new CppFunctionPointer(&default_f1))));
  return neuronPtr;
}

NeuronPtr
MLPfactory::makeNeuron(Handler Id, NeuronIteratorPtr neuronIteratorPtr,
    double totalAmountOfParameters)
{
  RNGScope scope;

  NeuronPtr neuronPtr(makeNeuron(Id));

  double extreme = sqrt(3 / totalAmountOfParameters);
  double weight;
  for (neuronIteratorPtr->first(); !neuronIteratorPtr->isDone(); neuronIteratorPtr->next())
    {
      weight = as<double> (runif(1, -extreme, extreme));
      neuronPtr->addCon(makeCon(*neuronIteratorPtr->currentItem(), weight));
    }

  MLPbehavior* mlpBehavior =
      dynamic_cast<MLPbehavior*> (neuronPtr->d_predictBehavior.get());
  mlpBehavior->d_bias = as<double> (runif(1, -extreme, extreme));

  return neuronPtr;
}

LayerPtr
MLPfactory::makeLayer()
{
  LayerPtr layerPtr(new SimpleContainer<NeuronPtr> );
  return layerPtr;
}

LayerContainerPtr
MLPfactory::makeLayerContainer()
{
  LayerContainerPtr layerContainerPtr(new SimpleContainer<LayerPtr> );
  layerContainerPtr->push_back(makeLayer());
  return layerContainerPtr;
}

NeuralNetworkPtr
MLPfactory::makeNeuralNetwork(NeuralFactory& neuralFactory)
{
  NeuralNetworkPtr neuralNetworkPtr(new SimpleNetwork(neuralFactory));
  return neuralNetworkPtr;
}



NeuralCreatorPtr
MLPfactory::makeNeuralCreator()
  {
    NeuralCreatorPtr neuralCreatorPtr(new SimpleNeuralCreator );
    return neuralCreatorPtr;
  }


ActivationFunctionPtr
MLPfactory::makeActivationFunction(NeuronPtr neuronPtr,
    Rcpp::XPtr<CppFunctionPointer> f0, Rcpp::XPtr<CppFunctionPointer> f1)
{
  ActivationFunctionPtr activationFunctionPtr(
      new ActivationFunction(neuronPtr, f0, f1));
  return activationFunctionPtr;
}

Rcpp::List
MLPfactory::makeXPtrFunctionList(std::string functionName)
{

  if (functionName == "Tanh")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Tanh_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Tanh_f1)));
    }
  else if (functionName == "Identity")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Identity_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Identity_f1)));
    }
  else if (functionName == "Threshold")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Threshold_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Threshold_f1)));
    }
  else if (functionName == "Logistic")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Logistic_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Logistic_f1)));
    }
  else if (functionName == "Exponential")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Exponential_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Exponential_f1)));
    }
  else if (functionName == "Reciprocal")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Reciprocal_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Reciprocal_f1)));
    }
  else if (functionName == "Square")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Square_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Square_f1)));
    }
  else if (functionName == "Gauss")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Gauss_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Gauss_f1)));
    }
  else if (functionName == "Sine")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Sine_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Sine_f1)));
    }
  else if (functionName == "Cosine")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Cosine_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Cosine_f1)));
    }
  else if (functionName == "Elliot")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Elliot_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Elliot_f1)));
    }
  else if (functionName == "Arctan")
    {
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Arctan_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&Arctan_f1)));
    }
  else
    {
      Rprintf(
          "\nWarning: default function applied. Unknown function name entered.\n");
      return List::create(
          _["f0"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&default_f0)),
          _["f1"] = XPtr<CppFunctionPointer> (
              new CppFunctionPointer(&default_f1)));
    }

}

File: pkg/Java/AmoreJ/javasrc/es/unileon/amorej/MLPFactory.java
/**
 * 
 */
package es.unileon.amorej;

import es.unileon.amorej.net.NeuralCreator;
import es.unileon.amorej.net.NeuralNetwork;
import es.unileon.amorej.net.SimpleNetwork;
import es.unileon.amorej.net.SimpleNeuralCreator;

/**
 * @author mcasl
 * 
 */
public class MLPFactory implements NeuralFactory {

	/*
	 * (non-Javadoc)
	 * 
	 * @see es.unileon.amorej.NeuralFactory#makeCon(es.unileon.amorej.Neuron,
	 * double)
	 */
	@Override
	public Connection makeConnection(Neuron neuron, double weight) {
		Connection connection = new Connection(neuron, weight);
		return connection;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see es.unileon.amorej.NeuralFactory#makeNeuron(java.lang.String)
	 */
	@Override
	public Neuron makeNeuron(String id, NeuralNetwork neuralNetwork) {
		Neuron neuron = new SimpleNeuron(neuralNetwork);
		neuron.setId(id);
		// TODO PredictBehaviorPtr
		// TODO predictBehaviorPtr(makePredictBehavior(neuronPtr));
		// TODO neuronPtr->setPredictBehavior(predictBehaviorPtr);
		// TODO ActivationFunctionPtr activationFunctionPtr(
		// TODO makeActivationFunction(neuronPtr,
		// TODO  Rcpp::XPtr<CppFunctionPointer>(new CppFunctionPointer(&default_f0)),
		// TODO  Rcpp::XPtr<CppFunctionPointer>(new CppFunctionPointer(&default_f1))));
		// TODO  neuronPtr->setActivationFunction(activationFunctionPtr);
		// TODO  NeuronTrainBehaviorPtr neuronTrainBehaviorPtr(
		// TODO makeHiddenNeuronTrainBehavior(neuronPtr));
		// TODO neuronPtr->setNeuronTrainBehavior(neuronTrainBehaviorPtr);
		return neuron;
	}

	public Container<Connection> makeConnectionContainer() {
		Container<Connection> connectionContainer = new SimpleContainer<Connection>();
		return connectionContainer;
	}

	public Container<Connection> makeConnectionContainer(int initialCapacity) {
		Container<Connection> connectionContainer = new SimpleContainer<Connection>(
				initialCapacity);
		return connectionContainer;
	}

	//TODO  PredictBehaviorPtr
	// MLPfactory::makePredictBehavior(NeuronPtr& neuronPtr)
	// {
	// PredictBehaviorPtr predictBehaviorPtr(new MLPbehavior(neuronPtr));
	// return predictBehaviorPtr;
	// }
	//

public 	Neuron makeNeuron(String Id, Iterator<Neuron> neuronIterator,
	 int totalAmountOfParameters, NeuralNetwork neuralNetwork) throws AmoreJException
	 {
	
	 Neuron neuron= this.makeNeuron(Id, neuralNetwork);	
	 double extreme = Math.sqrt(3.0 / totalAmountOfParameters);
	 double weight;
	 for (neuronIterator.first(); !neuronIterator.isDone(); neuronIterator.next()) {
		 weight = (Math.random()-0.5)*2*extreme;
		 Connection connection = this.makeConnection(neuronIterator.currentItem(), weight);
		 neuron.addConnection(connection);
	 }
// TODO	 MLPbehavior* mlpBehavior =	 dynamic_cast<MLPbehavior*> (neuronPtr->d_predictBehavior.get());
// TODO  mlpBehavior->d_bias = as<double> (runif(1, -extreme, extreme));
	 return neuron;
	 }
	

	public Container<Neuron> makeLayer() {
		Container<Neuron> layer = new SimpleContainer<Neuron>();
		return layer;
	}

	public Container<Container<Neuron>> makeLayerContainer()	 {
		Container<Container<Neuron>> layerContainer = new SimpleContainer<Container<Neuron>> ();
	 layerContainer.add( this.makeLayer() );
	 return layerContainer;
	 }

	public NeuralNetwork makeNeuralNetwork(NeuralFactory neuralFactory) {
		NeuralNetwork neuralNetwork = new SimpleNetwork(neuralFactory);
		// TODO NetworkTrainBehavior networkTrainBehavior(
		// TODO neuralFactory.makeNetworkTrainBehavior(neuralNetwork));
		// TODO  neuralNetwork.setNetworkTrainBehavior(networkTrainBehavior);
		// TODO  Remove comments above
		return neuralNetwork;
	}

	public NeuralCreator makeNeuralCreator() {
		NeuralCreator neuralCreator = new SimpleNeuralCreator();
		return neuralCreator;
	}

	//
	//TODO  ActivationFunctionPtr MLPfactory::makeActivationFunction(NeuronPtr& neuronPtr, Rcpp::XPtr<CppFunctionPointer> f0, Rcpp::XPtr<CppFunctionPointer> f1)
	// {
	// ActivationFunctionPtr activationFunctionPtr(
	// new ActivationFunction(neuronPtr, f0, f1));
	// return activationFunctionPtr;
	// }
	//
	//TODO  Rcpp::List
	// MLPfactory::makeXPtrFunctionList(std::string functionName)
	// {
	//
	// if (functionName == "Tanh")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Tanh_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Tanh_f1)));
	// }
	// else if (functionName == "Identity")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Identity_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Identity_f1)));
	// }
	// else if (functionName == "Threshold")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Threshold_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Threshold_f1)));
	// }
	// else if (functionName == "Logistic")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Logistic_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Logistic_f1)));
	// }
	// else if (functionName == "Exponential")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Exponential_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Exponential_f1)));
	// }
	// else if (functionName == "Reciprocal")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Reciprocal_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Reciprocal_f1)));
	// }
	// else if (functionName == "Square")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Square_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Square_f1)));
	// }
	// else if (functionName == "Gauss")
	// {
	// return List::create(
	// _["f0"]
	// = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Gauss_f0)),
	// _["f1"]
	// = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Gauss_f1)));
	// }
	// else if (functionName == "Sine")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Sine_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (new CppFunctionPointer(&Sine_f1)));
	// }
	// else if (functionName == "Cosine")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Cosine_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Cosine_f1)));
	// }
	// else if (functionName == "Elliot")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Elliot_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Elliot_f1)));
	// }
	// else if (functionName == "Arctan")
	// {
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Arctan_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&Arctan_f1)));
	// }
	// else
	// {
	// Rprintf(
	// "\nWarning: default function applied. Unknown function name entered.\n");
	// return List::create(
	// _["f0"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&default_f0)),
	// _["f1"] = XPtr<CppFunctionPointer> (
	// new CppFunctionPointer(&default_f1)));
	// }
	//
	// }
	//
	// TODO CostFunctionPtr
	// MLPfactory::makeCostFunction(std::string functionName)
	// {
	//
	// if (functionName == "LMS")
	// {
	// CostFunctionPtr costFunctionPtr(new LMS);
	// return costFunctionPtr;
	// }
	// else if (functionName == "LMLS")
	// {
	// CostFunctionPtr costFunctionPtr(new LMLS);
	// return costFunctionPtr;
	// }
	// else if (functionName == "TAO")
	// {
	// CostFunctionPtr costFunctionPtr(new Tao);
	// return costFunctionPtr;
	// }
	// else
	// {
	// throw std::invalid_argument(
	// "[SimpleNetwork::train Error]: Unknown cost function.");
	// }

	// }

}
